docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,197 @@
+Custom Attestation Data Formats for Open Enclave V3 Updates
+================
+
+As more custom attestation plugins are developed and more features are requested, new APIs are added to:
+
+1. Query registered attesters and verifiers.
+2. Select a format of the evidence that the registered attesters can generate.
+
+The following API was affected:
+
+1. Built-in SGX attester and verifier plugin
+
+This document describes the added and affected APIs as well as keeping
+the existing support for version 2 of the API.  This document assumes the user is familiar with
+the SGX attestation and [custom attestation](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation.md) with plugins.
+
+Motivation
+----------
+
+New APIs were added to help customers who would like to extend Open Enclave's current attestation structure with extra information while keep TEE-specific information within attester and verifier plugins.
+
+SGX report (referred as local attestation in Open Enclave) and SGX quote (referred as remote attestation in Open Enclave) generation takes different inputs and generates data of different structures. In order to simplify evidence generation in attestation applications, change was made to split built-in SGX attester into 2 attesters - one for SGX report and the other for SGX quote. Similar plugin splitting applies to built-in SGX verifier.
+
+Local and remote attestation is a TEE-specific feature. Specifying LOCAL or REMOTE flag when generating evidence can be confusing to many attesters. These flags were deprecated.
+
+User Experience
+---------------
+
+The user experience is improved by the ability of querying what the registered attestation verifiers can verify and attesters can attest. This is a typical attestation scenario:
+
+- Verifier application
+    1. Registers appropriate verifier plugins (oe_register_verifier).
+    2. Gets the format IDs of registered verifier plugins (oe_get_registered_verifier_format_ids).
+    3. Sends the format IDs to the attester application. These are the evidence formats that the verifier application is capable of verifying.
+- Attester application
+    1. Receives format IDs from verifier application.
+    2. Registers appropriate attester plugins for the platform so the evidence format selection in the next step can succeed (oe_register_attester).
```

Comment:
Upon initialization, an attester should be able to (self-)register all its supported plugins. This happens before it receives any format IDs from any verifiers.

Reply:
Right, that's another possible flow. The API design support many possible protocols.

---
In reply to: [387204152](https://github.com/openenclave/openenclave/pull/2621#discussion_r387204152) [](ancestors = 387204152)

Reply:
Which of the possible protocols are required to be supported by all TEEs, so that application software that implements this required protocol is guaranteed to work in all TEEs? One concern with the many possibilities is the potentially added confusion to plugin and application software developers, thus making the SDK less TEE agnostic.

Reply:
Offline discussion - the application has to explicitly register attesters and verifiers.

---
In reply to: [387207377](https://github.com/openenclave/openenclave/pull/2621#discussion_r387207377) [](ancestors = 387207377,387204152)

Reply:
Offline discussion - the application has to explicitly register attesters and verifiers. Except that, application code is TEE agnostic. See sample code at the bottom of this doc.

---
In reply to: [387233245](https://github.com/openenclave/openenclave/pull/2621#discussion_r387233245) [](ancestors = 387233245)